src/lint.cpp

```diff
@@ -500,11 +519,21 @@ void linter::propagate_variable_uses_to_parent_scope(
                var.name.normalized_name();
   };
 
+  bool has_eval = this->is_eval_used();
   for (const used_variable &used_var : current_scope.variables_used) {
     QLJS_ASSERT(!current_scope.declared_variables.find(used_var.name));
     const declared_variable *var =
         parent_scope.declared_variables.find(used_var.name);
-    if (var) {
+    if (used_var.name.normalized_name() == u8"eval") {
+      if (propagate_eval_use) {
+        (allow_variable_use_before_declaration
```

Comment:
I think `allow_variable_use_before_declaration` is always `false` if `propagate_eval_use` is `true`. Do you think that is significant? Should we assert it?
```suggestion
        QLJS_ASSERT(allow_variable_use_before_declaration);
        (allow_variable_use_before_declaration
```

If we assert that `allow_variable_use_before_declaration` is always `true` when propagating `eval`, then I think we can delete the changes below in the `current_scope.variables_used_in_descendant_scope` loop (lines 560-569). But maybe not. (The logic is complicated. xD)

Reply:
> QLJS_ASSERT(allow_variable_use_before_declaration);

did you mean this -
```suggestion
        (!allow_variable_use_before_declaration);
```

Reply:
> I think `allow_variable_use_before_declaration` is always `false` if `propagate_eval_use` is `true`. Do you think that is significant? Should we assert it?

For `visit_enter_function_scope_body` this doesn't follow now, I realized this after writing the test for that condition.